teamcity: failed tests on master: testrace/TestRangeSplitsWithConcurrentTxns

The following tests appear to have failed:

[#159617](https://teamcity.cockroachdb.com/viewLog.html?buildId=159617):

```
--- FAIL: testrace/TestRangeSplitsWithConcurrentTxns (0.590s)
Race detected!
testing.go:610: race detected during execution of test
------- Stdout: -------
I170223 14:57:28.157878 600 storage/store.go:1276  [n1,s1]: failed initial metrics computation: [n1,s1]: system config not yet available
I170223 14:57:28.158064 600 gossip/gossip.go:293  [n1] NodeDescriptor set to node_id:1 address:<network_field:"" address_field:"" > attrs:<> locality:<> 
I170223 14:57:28.163894 600 kv/split_test.go:157  starting split at key "G"...
I170223 14:57:28.177668 600 storage/replica_command.go:2397  [n1,s1,r1/1:/M{in-ax},@c4200c0900] initiating a split of this range at key "G" [r2]
I170223 14:57:28.223147 600 kv/split_test.go:161  split at key "G" complete
I170223 14:57:28.272119 600 kv/split_test.go:157  starting split at key "R"...
I170223 14:57:28.272700 600 storage/replica_command.go:2397  [n1,s1,r2/1:{"G"-/Max},@c4202da900] initiating a split of this range at key "R" [r3]
I170223 14:57:28.329911 600 kv/split_test.go:161  split at key "R" complete
I170223 14:57:28.368263 600 kv/split_test.go:157  starting split at key "a"...
I170223 14:57:28.368808 600 storage/replica_command.go:2397  [n1,s1,r3/1:{"R"-/Max},@c420170000] initiating a split of this range at key "a" [r4]
I170223 14:57:28.412992 600 kv/split_test.go:161  split at key "a" complete
I170223 14:57:28.458017 600 kv/split_test.go:157  starting split at key "l"...
I170223 14:57:28.458554 600 storage/replica_command.go:2397  [n1,s1,r4/1:{"a"-/Max},@c420170900] initiating a split of this range at key "l" [r5]
==================
Write at 0x00c420170a30 by goroutine 31:
  github.com/cockroachdb/cockroach/pkg/storage.(*Replica).setDescWithoutProcessUpdate()
      /go/src/github.com/cockroachdb/cockroach/pkg/storage/replica.go:1188 +0x163
  github.com/cockroachdb/cockroach/pkg/storage.(*Store).SplitRange()
      /go/src/github.com/cockroachdb/cockroach/pkg/storage/store.go:1932 +0x7c8
  github.com/cockroachdb/cockroach/pkg/storage.splitPostApply()
      /go/src/github.com/cockroachdb/cockroach/pkg/storage/store.go:1864 +0x458
  github.com/cockroachdb/cockroach/pkg/storage.(*Replica).handleReplicatedEvalResult()
      /go/src/github.com/cockroachdb/cockroach/pkg/storage/replica_proposal.go:557 +0x143e
  github.com/cockroachdb/cockroach/pkg/storage.(*Replica).handleEvalResult()
      /go/src/github.com/cockroachdb/cockroach/pkg/storage/replica_proposal.go:754 +0x1ed
  github.com/cockroachdb/cockroach/pkg/storage.(*Replica).processRaftCommand()
      /go/src/github.com/cockroachdb/cockroach/pkg/storage/replica.go:3688 +0xa0a
  github.com/cockroachdb/cockroach/pkg/storage.(*Replica).handleRaftReadyRaftMuLocked()
      /go/src/github.com/cockroachdb/cockroach/pkg/storage/replica.go:2785 +0xa6f
  github.com/cockroachdb/cockroach/pkg/storage.(*Replica).handleRaftReady()
      /go/src/github.com/cockroachdb/cockroach/pkg/storage/replica.go:2608 +0xf4
  github.com/cockroachdb/cockroach/pkg/storage.(*Store).processReady()
      /go/src/github.com/cockroachdb/cockroach/pkg/storage/store.go:3403 +0x17d
  github.com/cockroachdb/cockroach/pkg/storage.(*raftScheduler).worker()
      /go/src/github.com/cockroachdb/cockroach/pkg/storage/scheduler.go:214 +0x341
  github.com/cockroachdb/cockroach/pkg/storage.(*raftScheduler).Start.func2()
      /go/src/github.com/cockroachdb/cockroach/pkg/storage/scheduler.go:165 +0x4a
  github.com/cockroachdb/cockroach/pkg/util/stop.(*Stopper).RunWorker.func1()
      /go/src/github.com/cockroachdb/cockroach/pkg/util/stop/stopper.go:209 +0x8a

Previous read at 0x00c420170a30 by goroutine 98:
  github.com/cockroachdb/cockroach/pkg/storage.(*Replica).requiresExpiringLease()
      /go/src/github.com/cockroachdb/cockroach/pkg/storage/replica_range_lease.go:378 +0xae
  github.com/cockroachdb/cockroach/pkg/storage.(*Replica).leaseGoodToGo()
      /go/src/github.com/cockroachdb/cockroach/pkg/storage/replica.go:922 +0x97
  github.com/cockroachdb/cockroach/pkg/storage.(*Replica).redirectOnOrAcquireLease()
      /go/src/github.com/cockroachdb/cockroach/pkg/storage/replica.go:961 +0xad
  github.com/cockroachdb/cockroach/pkg/storage.(*Replica).tryAddWriteCmd()
      /go/src/github.com/cockroachdb/cockroach/pkg/storage/replica.go:2122 +0x1143
  github.com/cockroachdb/cockroach/pkg/storage.(*Replica).addWriteCmd()
      /go/src/github.com/cockroachdb/cockroach/pkg/storage/replica.go:2010 +0xa8
  github.com/cockroachdb/cockroach/pkg/storage.(*Replica).Send()
      /go/src/github.com/cockroachdb/cockroach/pkg/storage/replica.go:1420 +0x282
  github.com/cockroachdb/cockroach/pkg/storage.(*Store).Send()
      /go/src/github.com/cockroachdb/cockroach/pkg/storage/store.go:2529 +0x7ce
  github.com/cockroachdb/cockroach/pkg/storage.(*Stores).Send()
      /go/src/github.com/cockroachdb/cockroach/pkg/storage/stores.go:187 +0x24a
  github.com/cockroachdb/cockroach/pkg/kv.(*senderTransport).SendNext()
      /go/src/github.com/cockroachdb/cockroach/pkg/kv/transport.go:312 +0x290
  github.com/cockroachdb/cockroach/pkg/kv.(*localTestClusterTransport).SendNext()
      /go/src/github.com/cockroachdb/cockroach/pkg/kv/local_test_cluster_util.go:48 +0x96
  github.com/cockroachdb/cockroach/pkg/kv.(*DistSender).sendToReplicas()
      /go/src/github.com/cockroachdb/cockroach/pkg/kv/dist_sender.go:1143 +0x63e
  github.com/cockroachdb/cockroach/pkg/kv.(*DistSender).sendRPC()
      /go/src/github.com/cockroachdb/cockroach/pkg/kv/dist_sender.go:405 +0x3b3
  github.com/cockroachdb/cockroach/pkg/kv.(*DistSender).sendSingleRange()
      /go/src/github.com/cockroachdb/cockroach/pkg/kv/dist_sender.go:474 +0x1a7
  github.com/cockroachdb/cockroach/pkg/kv.(*DistSender).sendPartialBatch()
      /go/src/github.com/cockroachdb/cockroach/pkg/kv/dist_sender.go:931 +0x51e
  github.com/cockroachdb/cockroach/pkg/kv.(*DistSender).divideAndSendBatchToRanges()
      /go/src/github.com/cockroachdb/cockroach/pkg/kv/dist_sender.go:810 +0xc03
  github.com/cockroachdb/cockroach/pkg/kv.(*DistSender).Send()
      /go/src/github.com/cockroachdb/cockroach/pkg/kv/dist_sender.go:625 +0x3c9
  github.com/cockroachdb/cockroach/pkg/kv.(*TxnCoordSender).Send()
      /go/src/github.com/cockroachdb/cockroach/pkg/kv/txn_coord_sender.go:422 +0x2fe
  github.com/cockroachdb/cockroach/pkg/internal/client.(*DB).send()
      /go/src/github.com/cockroachdb/cockroach/pkg/internal/client/db.go:517 +0x1fc
  github.com/cockroachdb/cockroach/pkg/internal/client.(*Txn).sendInternal()
      /go/src/github.com/cockroachdb/cockroach/pkg/internal/client/txn.go:581 +0x102
  github.com/cockroachdb/cockroach/pkg/internal/client.(*Txn).send()
      /go/src/github.com/cockroachdb/cockroach/pkg/internal/client/txn.go:730 +0x75b
  github.com/cockroachdb/cockroach/pkg/internal/client.(*Txn).(github.com/cockroachdb/cockroach/pkg/internal/client.send)-fm()
      /go/src/github.com/cockroachdb/cockroach/pkg/internal/client/txn.go:317 +0x7c
  github.com/cockroachdb/cockroach/pkg/internal/client.sendAndFill()
      /go/src/github.com/cockroachdb/cockroach/pkg/internal/client/db.go:429 +0x11f
  github.com/cockroachdb/cockroach/pkg/internal/client.(*Txn).Run()
      /go/src/github.com/cockroachdb/cockroach/pkg/internal/client/txn.go:317 +0xfd
  github.com/cockroachdb/cockroach/pkg/internal/client.(*Txn).Put()
      /go/src/github.com/cockroachdb/cockroach/pkg/internal/client/txn.go:198 +0xe7
  github.com/cockroachdb/cockroach/pkg/kv.startTestWriter.func2()
      /go/src/github.com/cockroachdb/cockroach/pkg/kv/split_test.go:83 +0x274
  github.com/cockroachdb/cockroach/pkg/internal/client.(*DB).Txn.func1()
      /go/src/github.com/cockroachdb/cockroach/pkg/internal/client/db.go:479 +0x46
  github.com/cockroachdb/cockroach/pkg/internal/client.(*Txn).Exec()
      /go/src/github.com/cockroachdb/cockroach/pkg/internal/client/txn.go:511 +0x1dd
  github.com/cockroachdb/cockroach/pkg/internal/client.(*DB).Txn()
      /go/src/github.com/cockroachdb/cockroach/pkg/internal/client/db.go:480 +0x2c7
  github.com/cockroachdb/cockroach/pkg/kv.startTestWriter()
      /go/src/github.com/cockroachdb/cockroach/pkg/kv/split_test.go:89 +0x3c0

Goroutine 31 (running) created at:
  github.com/cockroachdb/cockroach/pkg/util/stop.(*Stopper).RunWorker()
      /go/src/github.com/cockroachdb/cockroach/pkg/util/stop/stopper.go:210 +0x73
  github.com/cockroachdb/cockroach/pkg/storage.(*raftScheduler).Start()
      /go/src/github.com/cockroachdb/cockroach/pkg/storage/scheduler.go:166 +0x1dd
  github.com/cockroachdb/cockroach/pkg/storage.(*Store).processRaft()
      /go/src/github.com/cockroachdb/cockroach/pkg/storage/store.go:3460 +0xad
  github.com/cockroachdb/cockroach/pkg/storage.(*Store).Start()
      /go/src/github.com/cockroachdb/cockroach/pkg/storage/store.go:1174 +0x58a
  github.com/cockroachdb/cockroach/pkg/testutils/localtestcluster.(*LocalTestCluster).Start()
      /go/src/github.com/cockroachdb/cockroach/pkg/testutils/localtestcluster/local_test_cluster.go:153 +0x13d5
  github.com/cockroachdb/cockroach/pkg/kv.createTestDBWithContext()
      /go/src/github.com/cockroachdb/cockroach/pkg/kv/txn_coord_sender_test.go:85 +0x13a
  github.com/cockroachdb/cockroach/pkg/kv.createTestDB()
      /go/src/github.com/cockroachdb/cockroach/pkg/kv/txn_coord_sender_test.go:76 +0x188
  github.com/cockroachdb/cockroach/pkg/kv.TestRangeSplitsWithConcurrentTxns()
      /go/src/github.com/cockroachdb/cockroach/pkg/kv/split_test.go:132 +0x9a
  testing.tRunner()
      /usr/local/go/src/testing/testing.go:657 +0x107

Goroutine 98 (running) created at:
  github.com/cockroachdb/cockroach/pkg/kv.TestRangeSplitsWithConcurrentTxns()
      /go/src/github.com/cockroachdb/cockroach/pkg/kv/split_test.go:148 +0x526
  testing.tRunner()
      /usr/local/go/src/testing/testing.go:657 +0x107
==================
I170223 14:57:28.546383 600 kv/split_test.go:161  split at key "l" complete
I170223 14:57:28.546463 600 kv/split_test.go:157  starting split at key "s"...
I170223 14:57:28.553414 600 storage/replica_command.go:2397  [n1,s1,r5/1:{"l"-/Max},@c420c62000] initiating a split of this range at key "s" [r6]
I170223 14:57:28.631923 600 kv/split_test.go:161  split at key "s" complete
I170223 14:57:28.688518 600 util/stop/stopper.go:493  quiescing; tasks left:
2      storage/intent_resolver.go:284
W170223 14:57:28.689050 755 storage/replica.go:2234  [n1,s1,r4/1:"{a"-l"},@c420170900] shutdown cancellation after 0.0s of attempting command Noop, Noop, Noop, ResolveIntent ["dZKjdKnaKY",/Min), Noop
W170223 14:57:28.689636 755 storage/intent_resolver.go:352  [n1,s1,r6/1:{"s"-/Max}]: failed to resolve intents: result is ambiguous (server shutdown)
I170223 14:57:28.689759 600 util/stop/stopper.go:493  quiescing; tasks left:
1      storage/intent_resolver.go:284
W170223 14:57:28.690214 512 storage/replica.go:2234  [n1,s1,r5/1:"{l"-s"},@c420c62000] shutdown cancellation after 0.0s of attempting command GC ["oiQMHvuTvA","oiQMHvuTvA\x00")
W170223 14:57:28.690598 512 storage/intent_resolver.go:395  could not GC completed transaction anchored at "oiQMHvuTvA": result is ambiguous (server shutdown)

```
Please assign, take a look and update the issue accordingly.

--- a/pkg/storage/replica.go
+++ b/pkg/storage/replica.go
@@ -919,14 +919,14 @@ func newNotLeaseHolderError(
 // existing lease is valid and owned by the current store. This method should
 // not be called directly. Use redirectOnOrAcquireLease instead.
 func (r *Replica) leaseGoodToGo(ctx context.Context) (LeaseStatus, bool) {
-	if r.requiresExpiringLease() {
-		// Slow-path for expiration-based leases.
-		return LeaseStatus{}, false
-	}
-
 	timestamp := r.store.Clock().Now()
 	r.mu.RLock()
 	defer r.mu.RUnlock()
+
+	if r.requiresExpiringLeaseRLocked() {
+		// Slow-path for expiration-based leases.
+		return LeaseStatus{}, false
+	}
 
 	status := r.leaseStatus(r.mu.state.Lease, timestamp, r.mu.minLeaseProposedTS)
 	switch status.state {
@@ -1011,7 +1011,7 @@ func (r *Replica) redirectOnOrAcquireLease(ctx context.Context) (LeaseStatus, *r
 				// leases, the lease is in need of renewal, and there's not
 				// already an extension pending.
 				_, requestPending := r.mu.pendingLeaseRequest.RequestPending()
-				if !requestPending && r.requiresExpiringLease() {
+				if !requestPending && r.requiresExpiringLeaseRLocked() {
 					renewal := status.lease.Expiration.Add(-int64(r.store.cfg.RangeLeaseRenewalDuration), 0)
 					if !timestamp.Less(renewal) {
 						if log.V(2) {

--- a/pkg/storage/replica_range_lease.go
+++ b/pkg/storage/replica_range_lease.go
@@ -85,6 +85,8 @@ func (p *pendingLeaseRequest) RequestPending() (roachpb.Lease, bool) {
 // Note: Once this function gets a context to be used for cancellation, instead
 // of replica.store.Stopper().ShouldQuiesce(), care will be needed for cancelling
 // the Raft command, similar to replica.addWriteCmd.
+//
+// Requires repl.mu is exclusively locked.
 func (p *pendingLeaseRequest) InitOrJoinRequest(
 	ctx context.Context,
 	repl *Replica,
@@ -119,7 +121,7 @@ func (p *pendingLeaseRequest) InitOrJoinRequest(
 		ProposedTS: &now,
 	}
 
-	if repl.requiresExpiringLease() {
+	if repl.requiresExpiringLeaseRLocked() {
 		reqLease.Expiration = status.timestamp.Add(int64(repl.store.cfg.RangeLeaseActiveDuration), 0)
 	} else {
 		// Get the liveness for the next lease holder and set the epoch in the lease request.
@@ -368,12 +370,12 @@ func (r *Replica) leaseStatus(
 	return status
 }
 
-// requiresExpiringLease returns whether this range uses an
-// expiration-based lease; false if epoch-based. Ranges located before
-// or including the node liveness table must use expiration leases to
-// avoid circular dependencies on the node liveness table. The replica
-// mutex must be held.
-func (r *Replica) requiresExpiringLease() bool {
+// requiresExpiringLeaseRLocked returns whether this range uses an
+// expiration-based lease; false if epoch-based. Ranges located before or
+// including the node liveness table must use expiration leases to avoid
+// circular dependencies on the node liveness table. The replica mutex must be
+// held.
+func (r *Replica) requiresExpiringLeaseRLocked() bool {
 	return r.store.cfg.NodeLiveness == nil || !r.store.cfg.EnableEpochRangeLeases ||
 		r.mu.state.Desc.StartKey.Less(roachpb.RKey(keys.NodeLivenessKeyMax))
 }

